### ISSUE TITLE
Is Security/JSONLoad really addressing a threat?

### ISSUE THREAD

**Issue by akimd** [rubocop/rubocop#11342]
## Expected behavior

I think it is an error from Ruby to propose only parsing of JSON via strings.  A decent implementation ought to be able to parse from a file without loading the whole file in RAM.  That's really bad in the case of large files, and once you have a stream-based parser, using it on strings is straightforward.

I don't despair that some day Ruby will have such a JSON parser.

As a consequence, I think that `JSON.parse(path.read)` is exactly what we should not write, but really `JSON.load(path)`.  This is "future-proof".

## Actual behavior

Security/JSONLoad complains about `JSON.load(path)`.  But when I read the code to which the very documentation of Security/JSONLoad points, I see:

```ruby
               # File json/lib/json/common.rb, line 323
def load(source, proc = nil, options = {})
  opts = load_default_options.merge options
  if source.respond_to? :to_str
    source = source.to_str
  elsif source.respond_to? :to_io
    source = source.to_io.read
  elsif source.respond_to?(:read)
    source = source.read
  end
  if opts[:allow_blank] && (source.nil? || source.empty?)
    source = 'null'
  end
  result = parse(source, opts)
  recurse_proc(result, &proc) if proc
  result
end
```

So basically:
1. fetch the string (`source`) that we are going to parse
2. call `parse` with default options ([they are](https://ruby-doc.org/stdlib-2.7.0/libdoc/json/rdoc/JSON.html#attribute-c-load_default_options) `{max_nesting: false, allow_nan: true, allow_blank: true}`)
3. if `proc` is defined, use it.

So AFAICT, this is no different from `JSON.parse(path.read)`, which the cop suggests.  The only place where things would be different is if I pass some `proc`, but I don't.  And if I do pass a `proc`, I mean to pass one, so it would be weird from Security/JSONLoad to complain that I use one and suggest that I move to `JSON.parse`.

In fact, when I look at [the documentation of more recent versions of Ruby](https://ruby-doc.org/core-3.1.2/JSON.html#method-i-load), the big warning about `JSON.load` is now clearly bound to the use of `proc`, and to that only.

So I don't think Security/JSONLoad's advices are useful.  I cannot see the danger it's trying to protect us from.  (Maybe the implementation of JSON.load has changed since the creation of the cop, I don't know).

## Steps to reproduce the problem
```
$ cat json.rb
require 'json'
require 'pathname'

JSON.load(Pathname('/tmp/foo.json'))
```

```
Inspecting 1 file
C

Offenses:

json.rb:1:1: C: [Correctable] Style/FrozenStringLiteralComment: Missing frozen string literal comment.
require 'json'
^
json.rb:3:6: C: [Correctable] Security/JSONLoad: Prefer JSON.parse over JSON.load.
JSON.load(Pathname('/tmp/foo.json'))
     ^^^^

1 file inspected, 2 offenses detected, 2 offenses autocorrectable
```

## RuboCop version

```
$ bundle exec rubocop -V
1.41.1 (using Parser 3.1.3.0, rubocop-ast 1.24.0, running on ruby 3.1.3) [arm64-darwin21]
```

Cheers!

**Comment by vlad-pisanov** [comment]
I _suspect_ that the danger comes from `load_default_options` used in `load`. With `parse`, you have to explicitly provide parsing options, but with `load`, (potentially dangerous) global defaults are used to parse untrusted data:

```ruby
require 'set'
require 'json/add/set'

attack = '{"json_class":"Set","a":[0,1,2]}'

JSON.parse(attack)
# => {"json_class"=>"Set", "a"=>[0, 1, 2]}     # a regular hash 👍

JSON.load(attack)
# => #<Set: {0, 1, 2}>                         # a un-marshalled Set object 😨
```

It also seems to me like the JSON documentation is just mangled up. If you look at an [older version](https://ruby-doc.org/stdlib-2.3.1/libdoc/json/rdoc/JSON.html#method-i-load), it has the same warning paragraph; I think it accidentally got moved to the wrong section.

**Comment by akimd** [comment]
@vlad-pisanov Thanks a lot for your answer.

OMG, the documentation is lying :(  https://ruby-doc.org/stdlib-2.7.0/libdoc/json/rdoc/JSON.html#attribute-c-load_default_options states:

```
The global default options for the [#load](https://ruby-doc.org/stdlib-2.7.0/libdoc/json/rdoc/JSON.html#method-i-load) method:

:max_nesting: false
:allow_nan: true
:allow_blank: true
```

But in reality
```
$ irb2.5 
irb(main):001:0> require 'json'
=> true
irb(main):002:0> JSON.load_default_options
=> {:max_nesting=>false, :allow_nan=>true, :allow_blank=>true, :create_additions=>true}
```
and it's precisely `create_additions: true` that matters here.

The documentation is fixed in newer versions of Ruby.

Your example would be a nice addition to the documentation of this cop.

Thanks again.

**Comment by akimd** [comment]
(And, imho, it should promote `JSON.load_file` instead of `JSON.parse` only.)
